README.md
# PyNetTools

<p align="center"><img src="https://socialify.git.ci/tomkat-cr/pynettools/image?description=0&amp;font=Inter&amp;language=1&amp;name=1&amp;owner=1&amp;pattern=Plus&amp;stargazers=0&amp;theme=Light" alt="project-image"></p>

## Overview

This is a simple networking tool in python that allows users to read from an input file with hostname and MAC address and update the system's host file accordingly. It was built to discover, map and maintain IP-Hostname mapping in a local network.

## Functionality

1. Discover all the 256 devices in the current network segment: List all hostname, MAC and IP addresses for items in the input file.
2. List all hostname, MAC and IP addresses for items in the input file.
3. Update the content of the host file from gathered IPs in the input file.
4. Update the host file from gathered IPs in the input file.
5. Show the content of the host file.
6. Check if the host name exists in the host file.
7. Insert HOSTNAME[:IP] mappings into the host file.
8. Remove mapping for HOSTNAME from host file.

## Project Organisation

Here is a tree view of the files in the repository:

```
.
├─ src
│   ├── Platform.py
│   ├── PyNetTools.py
│   ├── Host.py
│   ├── pynettools_main.py
│   ├── send_email.py
│   └── init_parser.py
├─ recycle_hosts_bkp.sh
├─ pymenttools.sh
├─ send_email.sh
├─ crontab-example.txt
├─ hostname_mac-example.txt
├─ LICENSE
├─ CHANGELOG.md
├─ .gitignore
├─ .env-example
└─ README.md

```

## Scripts

1. recycle_hosts_bkp.sh: This is a shell script to clean /etc/hosts backup files.
2. send_email.sh: This is a shell script to send emails through the SMTP server.
3. pynettools.sh: This is a shell script that executes the PyNetTools main function with necessary parameters.

## Input file example

The input file in the format `hostname_mac.txt` should look like this:

```
MAC_ADDRESS1 HOSTNAME1
MAC_ADDRESS2 HOSTNAME2
```

The provided `hostname_mac-example.txt` template file can be used to generate it:

```bash
cp hostname_mac-example.txt hostname_mac.txt
vi hostname_mac.txt
```

## Testing & Execution

To test and run this repository, follow the below steps:

1. Install Python3.

2. Generate the `.env` file to send Email notifications:

```bash
cp .env-example .env
vi .env
```

Specifiy these parameters:

```bash
# To send emails, set these parameters:
SMTP_SERVER=SMTP_SERVER_value
SMTP_PORT=SMTP_PORT_value
SMTP_USER=SMTP_USER_value
SMTP_PASSWORD=SMTP_PASSWORD_value
SMTP_DEFAULT_SENDER=SMTP_DEFAULT_SENDER_value
# Recycle files older than: ${PAR_MTIME} days.
PAR_MTIME=5
# Exclude filenames having ${EXCLUDE_FILENAMES_WITH} in it.
EXCLUDE_FILENAMES_WITH=ocrcorp
```

3. Run the desired functions via Python CLI or execute the shell scripts.

```bash
# PyNetTools /etc/host file update
sh pynettools.sh -u

# PyNetTools /etc/host backup recycler
sh recycle_hosts_bkp.sh
```

4. Include it in the crontab to refresh the host file constantly

```bash
crontab -e
```

Add the instructions in the `crontab-example.txt` file.


## Contributions

Feel free to submit pull requests, create issues or spread the word.

## License

MIT


CHANGELOG.md
# CHANGELOG

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/) and [Keep a Changelog](http://keepachangelog.com/).



## Unreleased
---

### New

### Changes

### Fixes

### Breaks


## 0.1.4 (2023-07-01)
---

### Change
README.md file revised.
".env-example" revised.


## 0.1.3 (2023-06-18)
---

### New
send_email.py and send_email.sh created.


## 0.1.2 (2023-06-17)
---

### New
CHANGELOG.md file created.
"hostname_mac-example.txt" and "hostname_mac.txt" files moved to project root.
pynettools.sh and recycle_hosts_bkp.sh created.
crontab-example.txt created.

### Changes
"get_ip_from_mac_addr.py" renamed to "pynettools_main.py".

### Fix
self.platform.is_mac is a method, not a property.
mac_addr to lower case also when it's linux.
nmap () substitution is also made in linux.
Debug platform print moved to main().


## 0.1.1 (2023-06-16)
---

### New
/etc/hosts file batch update with fresh IPs with -u option (update_from_input_entries, populate_input_entries).


## 0.1.0 (2023-06-15)
---

### New
Initital development of a simple program to get the IP address from a MAC address (get_ip_from_mac_addr.py), then converted to the project PyNetTools.
Get the IP address from a MAC address using the "arp -a" command (get_ip_from_mac).
Run "nmap" command to discover all network devices, prior to run "arp -a" (run_nmap, get_start_ip).
Get the current host IP address (get_local_ip).
List all hostname, MAC and IP addresses with -l option (list_input_entries).
Parse command line arguments and command help (init_parser).
Hostfile management class.
Platform management class.


src/send_email.py
# send_email.py
# 2023-06-18 | CR

# https://realpython.com/python-send-email/

from os import environ
from os.path import basename
import smtplib
import ssl
import argparse
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
from email.mime.application import MIMEApplication
from email.utils import COMMASPACE, formatdate


def send_email(sender_email, receiver_email, subject, text,
               html, files=[], debug=None):
    smtp_server = environ.get('SMTP_SERVER')
    smtp_port = environ.get('SMTP_PORT')  # For starttls
    smtp_user = environ.get('SMTP_USER')
    smtp_password = environ.get('SMTP_PASSWORD')
    if sender_email is None or sender_email.strip() == '':
        sender_email = environ.get('SMTP_DEFAULT_SENDER')

    message = MIMEMultipart("alternative")
    message["Subject"] = subject
    message["From"] = sender_email
    message['To'] = COMMASPACE.join(receiver_email)
    message['Date'] = formatdate(localtime=True)

    # Turn these into plain/html MIMEText objects
    body_plain_text = MIMEText(text, "plain")
    if html:
        body_html = MIMEText(html, "html")

    # Add HTML/plain-text parts to MIMEMultipart message
    # The email client will try to render the last part first
    message.attach(body_plain_text)
    if html:
        message.attach(body_html)

    for f in files or []:
        with open(f, "rb") as fil:
            part = MIMEApplication(
                fil.read(),
                Name=basename(f)
            )
        # After the file is closed
        part['Content-Disposition'] = 'attachment; filename="%s"' % basename(f)
        message.attach(part)

    if debug:
        print(f'smtp_server: {smtp_server}')
        print(f'smtp_port: {smtp_port}')
        print(f'smtp_user: {smtp_user}')
        print(f'smtp_password: {"*" * len(smtp_password)}')
        # print(f'smtp_password: {smtp_password}')
        print(f'sender_email: {sender_email}')
        print(f'receiver_email: {receiver_email}')
        print(f'subject: {subject}')
        print(f'text: {text}')
        print(f'html: {html}')
        print(f'file_path: {files}')

    # Create a secure SSL context
    context = ssl.create_default_context()
    # Try to log in to smtp server and send email
    try:
        smtp = smtplib.SMTP(smtp_server, smtp_port)
        # smtp.ehlo()  # Can be omitted
        smtp.starttls(context=context)  # Secure the connection
        # smtp.ehlo()  # Can be omitted
        smtp.login(smtp_user, smtp_password)
    except Exception as e:
        # Print any error messages to stdout
        print(f'Send_Email ERROR (preparing phase): {e}')
        return False
    try:
        smtp.sendmail(sender_email, receiver_email, message.as_string())
    except Exception as e:
        # Print any error messages to stdout
        print(f'Send_Email ERROR (sending phase): {e}')
    finally:
        smtp.close()
    return True


def init_parser():
    parser = argparse.ArgumentParser(
        description='Send an email'
    )
    parser.add_argument(
        "-D", "--debug",
        action="store_true",
        required=False,
        help="Show debug information",
    )
    parser.add_argument(
        "-f", "--sender",
        default='',
        required=False,
        help="Sender email",
    )
    parser.add_argument(
        "-t", "--to",
        nargs='+',
        default=[],
        required=True,
        help="Destination emails (separated by blanks)",
    )
    parser.add_argument(
        "-s", "--subject",
        nargs='+',
        default=[],
        required=True,
        help="Email subject",
    )
    parser.add_argument(
        "-m", "--message",
        nargs='+',
        default=[],
        required=False,
        help="Message (plain text)",
    )
    parser.add_argument(
        "-a", "--attachments",
        nargs='+',
        default=[],
        required=False,
        help="Attachments (filespecs separated by blanks)",
    )
    return parser


def main():
    parser = init_parser()
    args = parser.parse_args()

    send_email(
        args.sender,
        args.to,
        ' '.join(args.subject),
        ' '.join(args.message),
        None,
        args.attachments,
        args.debug
    )
    return


if __name__ == "__main__":
    main()
